### PR TITLE
Document how our CI works

### PR DESCRIPTION
**Pull request by inteon**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -274,3 +274,11 @@ certificate.cert-manager.io/demo-certificate   True    demo-cert-tls  1m
...
 secret/demo-cert-tls                     kubernetes.io/tls                     3      1m
 ```
+
+## Continuous Integration
+
+This project uses GitHub Actions to run continuous integration tests.
+There are two required test workflows:
+- `run_unit_tests` - this runs automatically on every pull request
+- `run_e2e_tests` - this runs on a pull request when the `ok-to-test` label is added  
+**⚠️ IMPORTANT: a maintainer must add this label manually after verifying that the PR is non-malicious, ping a maintainer when your PR is ready**
```
Comment (wallrj) [inline README.md:284]: Is there something to prevent the PR author adding the ok-to-test label themselves?
I can't remember whether external contributors are allowed to add labels by default.

It looks like the `ok-to-test` label gets removed after the e2e tests have completed. 
I suggest explaining that the ok-to-test label needs to be added each time.

Comment (inteon) [reply README.md:284]: adding labels is a privileged operation, this kind of setup is based on other setups (like https://github.com/cert-manager/aws-privateca-issuer/blob/main/.github/workflows/on-safe-to-test-label.yml).

Comment (wallrj) [reply README.md:284]: I see: https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#applying-a-label
> Anyone with triage access to a repository can apply and dismiss labels.

Comment (wallrj) [reply README.md:284]: How does the authentication to Google work for external PR authors?

https://github.com/jetstack/google-cas-issuer/blob/58bedee1b7a99b52917b6aacc4fed3c4c6a3e6d3/.github/workflows/pr_e2e_tests.yml#L33-L37

Won't the ephemeral GITHUB_TOKEN used in the GitHub Actions workflow be associated with the external contributor?
And they won't have permission to impersonate `google-cas-issuer-e2e@jetstack-cas.iam.gserviceaccount.com`, surely?

Or is the GITHUB_TOKEN owned by the triage user who added the label?

Comment (wallrj) [reply README.md:284]: I guess the token belongs to which ever user triggered the "event" which is "labeled" in this case....although I can't find any documentation that spells that out.

Comment (inteon) [reply README.md:284]: `pull_request_target` runs the action on the target repo, that is why it is dangerous and requires this ok-to-test label.

Comment (wallrj) [reply README.md]: I don't understand that statement.
I don't understand why `pull_request_target` is necessary?

The https://github.com/jetstack/google-cas-issuer/blob/main/.github/workflows/pr_e2e_tests.yml  workflow used to use the `pull_request` event.

https://github.com/jetstack/google-cas-issuer/commit/83ab764bc9363e881d479d45d27fbb9633a80f4f

And in https://github.com/jetstack/google-cas-issuer/pull/105 you wrote:
> TLDR: using pull_request_target runs the action in context of the target instead of the fork, thus allowing access to secrets etc

But the pr_e2e_tests.yml  workflow doesn't use any of the secrets AFAICS:
 * https://github.com/jetstack/google-cas-issuer/settings/secrets/actions

So if an external contributor created a PR last month (before you made the recent changes) the E2E tests would presumably have run, but failed, due to a failure to authenticate to Google CAS because GITHUB_TOKEN contained a token of the external user, who is not permitted to impersonate the Google service account.

But a Jetstack person could have re-run the E2E tests by manually triggering the workflow for that branch.
Why was that not sufficient?

Comment (inteon) [reply README.md]: How do you re-run the workflow and make it such that it uses the GITHUB_TOKEN of the non-external user?
I think that previously any PR that originated from a fork instead of a local branch would have failed.

Comment (wallrj) [reply README.md]: It would use the GITHUB_TOKEN of the user that triggered the workflow.

So it may have been enough to add the workflow_dispatch event, so that one of us could trigger the workflow from the Actions tab.

But as regards the documentation, I think the comments in the workflow file should explain that `pull_request_target` is used purely so that the workflow has permission to remove the label upon completion.
It is not used to provide access to secrets, from forks, because that workflow does not require those secrets.

Explain in the comments (or the docs) why we are ignoring the GitHub documentation which says:

> This event allows your workflow to do things like label or comment on pull requests from forks. Avoid using this event if you need to build or run code from the pull request.

-- https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target


If I'm wrong about that I still ask you to explain (in comments) how all this works, because I honestly don't understand it.
